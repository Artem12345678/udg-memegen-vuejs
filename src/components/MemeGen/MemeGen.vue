<template>
  <div class="app">
    <div class="googleFontFix">Fix</div>

    <div class="app__toolbar">
      <div class="form-row">
        <b-input-group prepend="File" class="col-lg-4 app__control">
          <b-form-file
            @change="onFileChangeHandler"
            accept="image/jpeg,image/png,image/gif"
            placeholder="Select"
            name="file"
          />
        </b-input-group>

        <b-input-group prepend="Top Text" class="col-lg-4 app__control">
          <b-form-input v-model="topText" name="topText" />
        </b-input-group>

        <b-input-group prepend="Bottom Text" class="col-lg-4 app__control">
          <b-form-input v-model="bottomText" name="bottomText" />
        </b-input-group>
      </div>

      <div class="form-row">
        <b-input-group prepend="Font" class="col-lg-4 app__control">
          <b-form-select v-model="fontFamily" :options="fontFamilyOptions" name="fontFamily" />
        </b-input-group>

        <verte ref="verte" value="#f0f" @input="fontColor = $event" picker="square" model="rgb" />
        <div class="input-group col-lg-4 app__control">
          <div class="input-group-prepend">
            <label class="input-group-text" for="color">Color</label>
          </div>
          <input
            class="form-control"
            readonly
            type="text"
            @click="toggleColorPicker"
            :value="fontColor"
            name="fontColor"
          />
        </div>

        <b-input-group prepend="Font Size" class="col-lg-4 app__control">
          <b-form-input
            style="flex-grow: 1; width: auto;"
            type="range"
            min="15"
            max="50"
            v-model="fontSize"
            name="fontSize"
          />
        </b-input-group>
      </div>
    </div>

    <div class="app__canvas">
      <canvas ref="canvas" id="canvas" />
    </div>

    <div class="app__save">
      <b-input-group class="col-lg-6 app__control">
        <b-form-select v-model="mimeType" :options="mimeTypeOptions" name="mimeType" />

        <template v-slot:append>
          <b-button @click="onSaveHandler" name="save" variant="info">Save As</b-button>
        </template>
      </b-input-group>
    </div>

    <div class="app__results">
      <div class="row">
        <div class="col-md-4" v-for="(item, index) in results" :key="index">
          <div class="card mb-5">
            <img class="card-img-top" alt="A custom meme" :src="item.imageUrl" />
            <div class="card-body">
              <p class="card-text">Top Text: {{ item.topText }}</p>
              <p class="card-text">Bottom Text: {{ item.bottomText }}</p>
              <a :href="item.imageUrl" download="meme" class="btn btn-success">Download</a>
              <button @click="onDeleteHandler(index)" class="btn btn-danger">Delete</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'MemeGen',
  data() {
    return {
      topText: 'Start Of Debug',
      bottomText: 'End Of Debug',
      fontFamily: 'Arial',
      fontSize: '25',
      fontColor: 'rgb(255, 0, 255)',
      mimeType: 'image/jpeg',
      imageUrl: 'https://i.imgflip.com/2cp1.jpg',
      fontFamilyOptions: [
        { value: null, text: 'Choose...', disabled: true },
        { value: 'Arial', text: 'Arial' },
        { value: 'Comic Sans MS', text: 'Comic Sans MS' },
        { value: 'Pacifico', text: 'Pacifico' },
      ],
      mimeTypeOptions: [
        { value: null, text: 'Choose...', disabled: true },
        { value: 'image/jpeg', text: 'JPG' },
        { value: 'image/png', text: 'PNG' },
        { value: 'image/gif', text: 'GIF' },
      ],
      results: [],
    };
  },

  computed: {
    canvasElement() {
      return this.$refs.canvas;
    },

    canvasContext() {
      return this.canvasElement.getContext('2d');
    },
  },

  mounted() {
    this.render();
  },

  updated() {
    this.render();
  },

  methods: {
    toggleColorPicker() {
      this.$refs.verte.isMenuActive = !this.$refs.verte.isMenuActive;
    },

    onFileChangeHandler(event) {
      const reader = new FileReader();

      reader.onload = () => {
        this.imageUrl = reader.result;

        this.render();
      };

      reader.readAsDataURL(event.target.files[0]);
    },

    onSaveHandler(event) {
      event.preventDefault();

      let snapshot;

      if (this.mimeType === 'image/gif') {
        // eslint-disable-next-line no-undef
        const gif = new GIF({
          workers: 2,
          quality: 10,
        });

        gif.addFrame(this.canvasElement, { delay: 200 });

        gif.on('finished', blob => {
          const reader = new FileReader();

          reader.onload = () => {
            snapshot = reader.result;

            this.results.push({
              topText: this.topText,
              bottomText: this.bottomText,
              imageUrl: snapshot,
            });
          };

          reader.readAsDataURL(blob);
        });

        gif.render();
      } else {
        snapshot = this.canvasElement.toDataURL(this.mimeType);

        this.results.push({
          topText: this.topText,
          bottomText: this.bottomText,
          imageUrl: snapshot,
        });
      }
    },

    onDeleteHandler(index) {
      this.results.splice(index, 1);
    },

    render() {
      const image = new Image();

      image.addEventListener('load', () => {
        this.canvasElement.width = image.width;
        this.canvasElement.height = image.height;

        this.canvasContext.drawImage(image, 0, 0);

        this.canvasContext.textAlign = 'center';
        this.canvasContext.textBaseline = 'middle';
        this.canvasContext.font = `${this.fontSize}px ${this.fontFamily}`;
        this.canvasContext.fillStyle = this.fontColor;
        this.canvasContext.fillText(this.topText, this.canvasElement.width / 2, 50);
        this.canvasContext.fillText(
          this.bottomText,
          this.canvasElement.width / 2,
          this.canvasElement.height - 50,
        );
      });

      image.setAttribute('crossorigin', 'anonymous'); // tainted canvas fix
      image.setAttribute('src', this.imageUrl);
    },
  },
};
</script>

<style lang="scss">
#canvas {
  max-width: 100%;
  max-height: 100%;
}

.googleFontFix {
  font-family: 'Pacifico', cursive;

  position: absolute;

  visibility: hidden;
  clip: rect(0 0 0 0);

  width: 1px;
  height: 1px;
  margin: -1px;
}

.verte__guide {
  display: none !important;
}

.app {
  &__control {
    margin-top: 1.5rem;
  }

  &__canvas {
    display: flex;

    margin-top: 1rem;
    padding: 20px;

    justify-content: center;
    align-items: center;
  }

  &__save {
    display: flex;

    align-items: center;
    justify-content: center;
  }

  &__results {
    margin-top: 2.5rem;
  }
}
</style>
