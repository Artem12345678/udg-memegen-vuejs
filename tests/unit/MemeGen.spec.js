import { createLocalVue, mount } from '@vue/test-utils';
import { BootstrapVue } from 'bootstrap-vue';
import Verte from 'verte';

import MemeGen from '@/components/MemeGen/MemeGen.vue';

const localVue = createLocalVue();

localVue.use(BootstrapVue);

localVue.component('verte', Verte);

describe('MemeGen.vue', () => {
  let wrapper;

  const data = {
    topText: 'Start Of Debug',
    bottomText: 'End Of Debug',
    fontFamily: 'Arial',
    fontSize: '25',
    fontColor: 'rgb(255, 0, 255)',
    mimeType: 'image/jpeg',
    imageUrl: 'https://i.imgflip.com/2cp1.jpg',
    fontFamilyOptions: [
      { value: null, text: 'Choose...', disabled: true },
      { value: 'Arial', text: 'Arial' },
      { value: 'Comic Sans MS', text: 'Comic Sans MS' },
      { value: 'Pacifico', text: 'Pacifico' },
    ],
    mimeTypeOptions: [
      { value: null, text: 'Choose...', disabled: true },
      { value: 'image/jpeg', text: 'JPG' },
      { value: 'image/png', text: 'PNG' },
      { value: 'image/gif', text: 'GIF' },
    ],
    results: [],
  };

  const onFileChangeHandler = jest.fn();
  const onSaveHandler = jest.fn();
  const render = jest.fn();

  beforeEach(() => {
    wrapper = mount(MemeGen, {
      localVue,
      sync: false,
      methods: {
        onFileChangeHandler,
        onSaveHandler,
        render,
      },
    });

    wrapper.setData(data);
  });

  it('matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('returnes canvas element from canvasElement computed property', () => {
    expect(wrapper.vm.canvasElement).toBe(wrapper.find({ ref: 'canvas' }).element);
  });

  it('returnes 2d context of canvas element from canvasContext computed property', () => {
    expect(wrapper.vm.canvasContext.canvas).toBe(wrapper.find({ ref: 'canvas' }).element);
  });

  describe('File input element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="file"]').exists()).toBe(true);
    });

    it('triggers onFileChange() method', () => {
      wrapper.find('[name="file"]').trigger('change');

      expect(onFileChangeHandler).toBeCalled();
    });
  });

  describe('Top Text input element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="topText"]').exists()).toBe(true);
    });

    it('has right default value', () => {
      expect(wrapper.find('[name="topText"]').element.value).toBe(data.topText);
    });

    it('updates corresponding model property', () => {
      const newValue = 'Top Text Test';

      wrapper.find('[name="topText"]').setValue(newValue);

      expect(wrapper.vm.$data.topText).toBe(newValue);
      expect(wrapper.find('[name="topText"]').element.value).toBe(newValue);
    });
  });

  describe('Bottom Text input element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="bottomText"]').exists()).toBe(true);
    });

    it('has right default value', () => {
      expect(wrapper.find('[name="bottomText"]').element.value).toBe(data.bottomText);
    });

    it('updates corresponding model property', () => {
      const newValue = 'Bottom Text Test';

      wrapper.find('[name="bottomText"]').setValue(newValue);

      expect(wrapper.vm.$data.bottomText).toBe(newValue);
      expect(wrapper.find('[name="bottomText"]').element.value).toBe(newValue);
    });
  });

  describe('Font Size input element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="fontSize"]').exists()).toBe(true);
    });

    it('has right default value', () => {
      expect(wrapper.find('[name="fontSize"]').element.value).toBe(data.fontSize);
    });

    it('updates corresponding model property', () => {
      const newValue = '50';

      wrapper.find('[name="fontSize"]').setValue(newValue);

      expect(wrapper.vm.$data.fontSize).toBe(newValue);
      expect(wrapper.find('[name="fontSize"]').element.value).toBe(newValue);
    });
  });

  describe('Font Family select element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="fontFamily"]').exists()).toBe(true);
    });

    it('renders the right amount of options', () => {
      expect(wrapper.find('[name="fontFamily"]').findAll('option').length).toBe(
        data.fontFamilyOptions.length,
      );
    });

    it('has right default value', () => {
      expect(wrapper.find('[name="fontFamily"]').element.value).toBe(data.fontFamily);
    });

    it('updates corresponding model property', async () => {
      const newValue = 'Pacifico';

      wrapper
        .find('[name="fontFamily"]')
        .findAll('option')
        .at(3)
        .setSelected();

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.fontFamily).toBe(newValue);
      expect(wrapper.find('[name="fontFamily"]').element.value).toBe(newValue);
    });
  });

  describe('Mime Type select element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="mimeType"]').exists()).toBe(true);
    });

    it('renders the right amount of options', () => {
      expect(wrapper.find('[name="mimeType"]').findAll('option').length).toBe(
        data.mimeTypeOptions.length,
      );
    });

    it('has right default value', () => {
      expect(wrapper.find('[name="mimeType"]').element.value).toBe(data.mimeType);
    });

    it('updates corresponding model property', async () => {
      const newValue = 'image/gif';

      wrapper
        .find('[name="mimeType"]')
        .findAll('option')
        .at(3)
        .setSelected();

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.mimeType).toBe(newValue);
      expect(wrapper.find('[name="mimeType"]').element.value).toBe(newValue);
    });
  });

  describe('Colorpicker element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="fontColor"]').exists()).toBe(true);
      expect(wrapper.find({ ref: 'verte' }).exists()).toBe(true);
    });

    it('shows up on click', () => {
      const previousState = wrapper.find({ ref: 'verte' }).vm.isMenuActive;

      wrapper.find('[name="fontColor"]').trigger('click');

      expect(wrapper.find({ ref: 'verte' }).vm.isMenuActive).toBe(!previousState);
    });

    it('updates corresponding model property', async () => {
      const newValue = 'rgb(255,255,255)';

      wrapper.find({ ref: 'verte' }).vm.$emit('input', newValue);

      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.fontColor).toBe(newValue);
      expect(wrapper.find('[name="fontColor"]').element.value).toBe(newValue);
    });
  });

  describe('Save As button element', () => {
    it('exists', () => {
      expect(wrapper.find('[name="save"]').exists()).toBe(true);
    });

    it('triggers onSaveHandler() method', () => {
      wrapper.find('[name="save"]').trigger('click');

      expect(onSaveHandler).toBeCalled();
    });
  });

  describe('Results section', () => {
    beforeEach(() => {
      wrapper.setData({
        results: [
          {
            topText: 'Top Text Test',
            bottomText: 'Bottom Text Test',
            imageUrl: 'imageUrl',
          },
        ],
      });
    });

    it('exists', () => {
      expect(wrapper.find('.app__results .row').exists()).toBe(true);
    });

    it('renders the right amount of elements', () => {
      expect(wrapper.findAll('.app__results .row .col-md-4').length).toBe(
        wrapper.vm.$data.results.length,
      );
    });

    it('renders right Top Text string', () => {
      expect(
        wrapper
          .findAll('.app__results .row .col-md-4 p.card-text')
          .at(0)
          .text(),
      ).toBe(`Top Text: ${wrapper.vm.$data.results[0].topText}`);
    });

    it('renders right Bottom Text string', () => {
      expect(
        wrapper
          .findAll('.app__results .row .col-md-4 p.card-text')
          .at(1)
          .text(),
      ).toBe(`Bottom Text: ${wrapper.vm.$data.results[0].bottomText}`);
    });

    it('renders right href attribute for the Download link element', () => {
      expect(wrapper.find('.app__results .row .col-md-4 .btn-success').attributes('href')).toBe(
        wrapper.vm.$data.results[0].imageUrl,
      );
    });

    describe('Delete button element', () => {
      it('exists', () => {
        expect(wrapper.find('.app__results .row .col-md-4 .btn-danger').exists()).toBe(true);
      });

      it('deletes corresponding element on click', async () => {
        wrapper.find('.app__results .row .col-md-4 .btn-danger').trigger('click', 0);

        await wrapper.vm.$nextTick();

        expect(wrapper.findAll('.app__results .row .col-md-4 .btn-danger').length).toBe(
          wrapper.vm.$data.results.length,
        );
      });
    });
  });
});
