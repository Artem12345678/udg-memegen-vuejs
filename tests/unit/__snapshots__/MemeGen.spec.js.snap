// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemeGen.vue matches snapshot 1`] = `
<div
  class="app"
>
  <div
    class="googleFontFix"
  >
    Fix
  </div>
   
  <div
    class="app__toolbar"
  >
    <div
      class="form-row"
    >
      <div
        class="input-group col-lg-4 app__control"
        role="group"
      >
        <div
          class="input-group-prepend"
        >
          <div
            class="input-group-text"
          >
            File
          </div>
          <!---->
        </div>
        <div
          class="custom-file b-form-file"
          id="__BVID__4__BV_file_outer_"
        >
          <input
            accept="image/jpeg,image/png,image/gif"
            class="custom-file-input"
            id="__BVID__4"
            name="file"
            type="file"
          />
          <label
            class="custom-file-label"
            data-browse="Browse"
            for="__BVID__4"
          >
            Select
          </label>
        </div>
        <!---->
      </div>
       
      <div
        class="input-group col-lg-4 app__control"
        role="group"
      >
        <div
          class="input-group-prepend"
        >
          <div
            class="input-group-text"
          >
            Top Text
          </div>
          <!---->
        </div>
        <input
          class="form-control"
          id="__BVID__5"
          name="topText"
          type="text"
        />
        <!---->
      </div>
       
      <div
        class="input-group col-lg-4 app__control"
        role="group"
      >
        <div
          class="input-group-prepend"
        >
          <div
            class="input-group-text"
          >
            Bottom Text
          </div>
          <!---->
        </div>
        <input
          class="form-control"
          id="__BVID__6"
          name="bottomText"
          type="text"
        />
        <!---->
      </div>
    </div>
     
    <div
      class="form-row"
    >
      <div
        class="input-group col-lg-4 app__control"
        role="group"
      >
        <div
          class="input-group-prepend"
        >
          <div
            class="input-group-text"
          >
            Font
          </div>
          <!---->
        </div>
        <select
          class="custom-select"
          id="__BVID__7"
          name="fontFamily"
        >
          <option
            disabled="disabled"
            value=""
          >
            Choose...
          </option>
          <option
            value="Arial"
          >
            Arial
          </option>
          <option
            value="Comic Sans MS"
          >
            Comic Sans MS
          </option>
          <option
            value="Pacifico"
          >
            Pacifico
          </option>
        </select>
        <!---->
      </div>
       
      <div
        class="verte"
      >
        <button
          class="verte__guide"
          style="color: rgb(255, 0, 255); fill: rgb(255,0,255);"
          type="button"
        >
          <svg
            class="verte__icon"
            viewBox="0 0 24 24"
          >
            <pattern
              fill="FFF"
              height="6"
              id="checkerboard"
              patternUnits="userSpaceOnUse"
              width="6"
            >
              <rect
                fill="#7080707f"
                height="3"
                width="3"
                x="0"
                y="0"
              />
              <rect
                fill="#7080707f"
                height="3"
                width="3"
                x="3"
                y="3"
              />
            </pattern>
            <circle
              cx="12"
              cy="12"
              fill="url(#checkerboard)"
              r="12"
            />
            <circle
              cx="12"
              cy="12"
              r="12"
            />
          </svg>
        </button>
        <div
          class="verte__menu-origin verte__menu-origin--bottom"
        >
          <div
            class="verte__menu"
            style="transform: translate(0px, 0px);"
            tabindex="-1"
          >
            <button
              class="verte__close"
              type="button"
            >
              <svg
                class="verte__icon verte__icon--small"
                viewBox="0 0 24 24"
              >
                <title>
                  Close Icon
                </title>
                <path
                  d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
                />
              </svg>
            </button>
            <div
              class="verte__draggable"
            />
            <div
              class="verte-picker verte-picker--square"
            >
              <div
                class="verte-picker__origin"
              >
                <canvas
                  class="verte-picker__canvas"
                  height="150"
                  width="250"
                />
                <div
                  class="verte-picker__cursor"
                  style="transform: translate3d(0px, 0px, 0);"
                />
              </div>
              <div
                class="slider verte-picker__slider"
              >
                <div
                  class="slider__track"
                >
                  <div
                    class="slider__fill"
                  />
                  <div
                    class="slider__handle"
                    style="transform: translate(0px, 0); background-color: rgb(255, 0, 0);"
                  >
                    <!---->
                  </div>
                </div>
                <input
                  class="slider__input"
                  style="display: none;"
                  type="number"
                />
              </div>
              <!---->
            </div>
            <div
              class="verte__controller"
            >
              <div
                class="slider"
              >
                <div
                  class="slider__track"
                >
                  <div
                    class="slider__fill"
                  />
                  <div
                    class="slider__handle"
                    style="transform: translate(0px, 0); background-color: rgba(255, 0, 255, 0);"
                  >
                    <!---->
                  </div>
                </div>
                <input
                  class="slider__input"
                  style="display: none;"
                  type="number"
                />
              </div>
              <!---->
              <div
                class="verte__inputs"
              >
                <button
                  class="verte__model"
                  type="button"
                >
                  rgb
                </button>
                <!---->
                <input
                  class="verte__input"
                  max="255"
                  min="0"
                  type="number"
                />
                <input
                  class="verte__input"
                  max="255"
                  min="0"
                  type="number"
                />
                <input
                  class="verte__input"
                  max="255"
                  min="0"
                  type="number"
                />
                <!---->
                <button
                  class="verte__submit"
                  type="button"
                >
                  <title>
                    Submit Icon
                  </title>
                  <svg
                    class="verte__icon"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="verte__recent"
              >
                <a
                  class="verte__recent-color"
                  href="#"
                  role="button"
                  style="color: rgb(8, 43, 100);"
                />
                <a
                  class="verte__recent-color"
                  href="#"
                  role="button"
                  style="color: rgb(71, 173, 229);"
                />
                <a
                  class="verte__recent-color"
                  href="#"
                  role="button"
                  style="color: rgb(59, 85, 255);"
                />
                <a
                  class="verte__recent-color"
                  href="#"
                  role="button"
                  style="color: rgb(59, 210, 161);"
                />
                <a
                  class="verte__recent-color"
                  href="#"
                  role="button"
                  style="color: rgb(110, 139, 218);"
                />
                <a
                  class="verte__recent-color"
                  href="#"
                  role="button"
                  style="color: rgb(255, 108, 46);"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="input-group col-lg-4 app__control"
      >
        <div
          class="input-group-prepend"
        >
          <label
            class="input-group-text"
            for="color"
          >
            Color
          </label>
        </div>
         
        <input
          class="form-control"
          name="fontColor"
          readonly="readonly"
          type="text"
        />
      </div>
       
      <div
        class="input-group col-lg-4 app__control"
        role="group"
      >
        <div
          class="input-group-prepend"
        >
          <div
            class="input-group-text"
          >
            Font Size
          </div>
          <!---->
        </div>
        <input
          class="custom-range"
          id="__BVID__13"
          max="50"
          min="15"
          name="fontSize"
          style="flex-grow: 1; width: auto;"
          type="range"
        />
        <!---->
      </div>
    </div>
  </div>
   
  <div
    class="app__canvas"
  >
    <canvas
      id="canvas"
    />
  </div>
   
  <div
    class="app__save"
  >
    <div
      class="input-group col-lg-6 app__control"
      role="group"
    >
      <!---->
      <select
        class="custom-select"
        id="__BVID__14"
        name="mimeType"
      >
        <option
          disabled="disabled"
          value=""
        >
          Choose...
        </option>
        <option
          value="image/jpeg"
        >
          JPG
        </option>
        <option
          value="image/png"
        >
          PNG
        </option>
        <option
          value="image/gif"
        >
          GIF
        </option>
      </select>
      <div
        class="input-group-append"
      >
        <!---->
        <button
          class="btn btn-info"
          name="save"
          type="button"
        >
          Save As
        </button>
      </div>
    </div>
  </div>
   
  <div
    class="app__results"
  >
    <div
      class="row"
    />
  </div>
</div>
`;
